elasticdl/python/elasticdl/local_executor.py

```diff
@@ -0,0 +1,182 @@
+
+_MockedTask = namedtuple("Task", ["shard_name", "start", "end"])
+
+
+class LocalExecutor:
+    """Train the Keras model defined in ElasticDL model zoo locally
+    using the custom training loop with Tensorflow native low-level API.
+    """
+
+    def __init__(self, args):
+        envs = parse_envs(args.envs)
+        self._init_environment(envs)
+
+        (
+            self.model_inst,
+            self.dataset_fn,
+            self.loss_fn,
+            self.opt_fn,
+            self.eval_metrics_fn,
+            self.prediction_outputs_processor,
+            self.custom_data_reader,
+        ) = get_model_spec(
+            model_zoo=args.model_zoo,
+            model_def=args.model_def,
+            dataset_fn=args.dataset_fn,
+            loss=args.loss,
+            optimizer=args.optimizer,
+            eval_metrics_fn=args.eval_metrics_fn,
+            model_params=args.model_params,
+            prediction_outputs_processor="",
+            custom_data_reader="",
+        )
+        self.opt = self.opt_fn()
+        self.epoch = args.num_epochs
+        self.evaluation_steps = args.evaluation_steps
+        self.batch_size = args.minibatch_size
+        self.data_reader_params = get_dict_from_params_str(
+            args.data_reader_params
+        )
+        self.records_per_task = (
+            args.minibatch_size * args.num_minibatches_per_task
+        )
+        self.data_reader = create_data_reader(
+            data_origin=args.training_data,
+            records_per_task=self.records_per_task,
+            **self.data_reader_params
+        )
+        self.training_data = args.training_data
+        self.validation_data = args.validation_data
+
+    def _init_environment(self, envs):
+        for key, value in envs.items():
+            os.environ[key] = value
+
+    def run(self):
+        """Execute the training loop"""
+        epoch = 0
+        step = 0
+
+        train_tasks = self._gen_tasks(self.training_data)
+        validation_tasks = self._gen_tasks(self.validation_data)
+        train_dataset = self._get_dataset(train_tasks)
+        validation_dataset = self._get_dataset(validation_tasks)
+
+        while epoch < self.epoch:
+            for features, labels in train_dataset:
+                loss = self._train(features, labels)
+                logger.info("step {}, Loss = {}".format(step, loss))
+                step += 1
+                if (
+                    self.evaluation_steps > 0
+                    and step % self.evaluation_steps == 0
+                ):
+                    self._evaluate(validation_dataset)
+            self._evaluate(validation_dataset)
+            logger.info("Epoch {} end".format(epoch))
+            epoch += 1
+
+    def _train(self, features, labels):
+        with tf.GradientTape() as tape:
+            outputs = self.model_inst.call(features, training=True)
+            loss = self.loss_fn(labels, outputs)
```

Comment:
Add regularization loss if it exists.
See [worker.py](https://github.com/sql-machine-learning/elasticdl/blob/645a1399efad2629d0a356ee3f5409dbfae3c6ea/elasticdl/python/worker/worker.py#L643)